pyro/infer/abstract_infer.py

```diff
@@ -71,3 +76,46 @@ def run(self, *args, **kwargs):
             self.exec_traces.append(tr)
             self.log_weights.append(logit)
         return self
+
+
+class PosteriorPredictive(TracePosterior):
+    """
+    Generates and holds traces from the posterior predictive distribution,
+    given model execution traces from the approximate posterior. This is
+    achieved by constraining latent sites to randomly sampled parameter
+    values from the model execution traces and running the model forward
+    to generate traces with new response ("_RETURN") sites.
+
+    :param model: arbitrary Python callable containing Pyro primitives.
+    :param model_traces: execution traces from the model.
+    :param int num_samples: number of samples to generate.
+    :param list hide_nodes: list of nodes that should be hidden when
+        replaying the model against ``model_traces``. Usually, this
+        corresponds to observed nodes.
+    :param list log_weights: optional importance weights of execution
+        traces to be used during sampling.
+    """
+    def __init__(self, model, model_traces, num_samples, hide_nodes=[], log_weights=None):
+        self.model = model
+        self.model_traces = model_traces
+        self.hide_nodes = hide_nodes
+        self.num_samples = num_samples
+        if not log_weights:
+            log_weights = torch.zeros(len(model_traces,))
+        else:
+            log_weights = torch.tensor(log_weights)
+        self._categorical = Categorical(logits=log_weights)
+        super(PosteriorPredictive, self).__init__()
+
+    def _get_random_pruned_trace(self):
+        random_idx = self._categorical.sample()
+        trace = self.model_traces[random_idx].copy()
+        for node in self.hide_nodes:
+            trace.remove_node(node)
+        return trace
+
+    def _traces(self, *args, **kwargs):
+        for _ in range(self.num_samples):
+            model_trace = self._get_random_pruned_trace()
+            replayed_trace = poutine.trace(poutine.replay(self.model, model_trace)).get_trace(*args, **kwargs)
```

Comment:
Passing this explicitly makes it easier to generate predictions on new data though. Otherwise, we would have to rewrite the same machinery separately.